plugin/src/main/java/org/opensearch/ml/action/deploy/TransportDeployModelAction.java

```diff
@@ -108,125 +116,143 @@ public TransportDeployModelAction(
         this.mlModelManager = mlModelManager;
         this.mlStats = mlStats;
         allowCustomDeploymentPlan = ML_COMMONS_ALLOW_CUSTOM_DEPLOYMENT_PLAN.get(settings);
+        filterByEnabled = ML_COMMONS_VALIDATE_BACKEND_ROLES.get(settings);
         clusterService
             .getClusterSettings()
             .addSettingsUpdateConsumer(ML_COMMONS_ALLOW_CUSTOM_DEPLOYMENT_PLAN, it -> allowCustomDeploymentPlan = it);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(ML_COMMONS_VALIDATE_BACKEND_ROLES, it -> filterByEnabled = it);
+
     }
 
     @Override
     protected void doExecute(Task task, ActionRequest request, ActionListener<MLDeployModelResponse> listener) {
         MLDeployModelRequest deployModelRequest = MLDeployModelRequest.fromActionRequest(request);
         String modelId = deployModelRequest.getModelId();
-        String[] targetNodeIds = deployModelRequest.getModelNodeIds();
-        boolean deployToAllNodes = targetNodeIds == null || targetNodeIds.length == 0;
-        if (!allowCustomDeploymentPlan && !deployToAllNodes) {
-            throw new IllegalArgumentException("Don't allow custom deployment plan");
-        }
+        User user = RestActionUtils.getUserContext(client);
+        String[] excludes = new String[] { MLModel.MODEL_CONTENT_FIELD, MLModel.OLD_MODEL_CONTENT_FIELD };
 
-        // mlStats.getStat(MLNodeLevelStat.ML_NODE_EXECUTING_TASK_COUNT).increment();
-        mlStats.getStat(MLNodeLevelStat.ML_NODE_TOTAL_REQUEST_COUNT).increment();
-        DiscoveryNode[] allEligibleNodes = nodeFilter.getEligibleNodes();
-        Map<String, DiscoveryNode> nodeMapping = new HashMap<>();
-        for (DiscoveryNode node : allEligibleNodes) {
-            nodeMapping.put(node.getId(), node);
-        }
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+            mlModelManager.getModel(modelId, null, excludes, ActionListener.wrap(mlModel -> {
+                SecurityUtils.validateModelGroupAccess(user, mlModel.getModelGroupId(), client, ActionListener.wrap(access -> {
+                    if ((filterByEnabled) && (Boolean.FALSE.equals(access))) {
```

Comment:
If we can make sure there's no null returned by `validateModelGroupAccess`, then we can use primitive boolean to save the boxing/unboxing overhead.

Reply:
Need to test few edge cases for this. Maybe using boxing/unboxing could be a safer option. If there is no case null is returned, will change them to primitive ones

Reply:
It would be better to think all the cases through before implementing, test may cover some cases but not all because some cases would be difficult to mimic. Better think through if we need to return null or not here.